clients/cppsal/include/sal_node.h

```diff
@@ -0,0 +1,658 @@
+#pragma once
+
+#include "sal_data.h"
+#include "sal_exception.h"
+#include <cassert>
+
+#include "Poco/DateTimeFormatter.h"
+
+namespace sal
+{
+    /// tree-node structure supports std::string path like in file system
+    /// is there any third-party library provide this data structure?
+    namespace node
+    {
+        typedef enum
+        {
+            NODE_BRANCH,
+            NODE_LEAF
+        } NodeType;
+
+
+        // TODO: those global names should be removed
+        // static char JSON_CONTENT_REPORT[] = "report"; // should be named as "summary"
+        // static char JSON_CONTENT_OBJECT[] = "object";
+
+        // static char OBJ_TYPE_FULL[] = "full";
+        // static char OBJ_TYPE_SUMMARY[] = "summary";
+
+        /// corresponding to python decorator `dataobject.register()`
+        /// each item in the Branch children list contains this information
+        /// Refactored: rename from LeafType to NodeInfo mimic FileInfo class
+        class NodeInfo
+        {
+
+        public:
+            // TODO: set a proper default values for members
+            NodeInfo()
+                    : version(SAL_API_VERSION){};
+            NodeInfo(const std::string _cls, const std::string _group, const uint64_t _version,
+                     const std::string _name = "")
+                    : cls(_cls)
+                    , group(_group)
+                    , version(_version)
+                    , name(_name){};
+
+            std::string cls;   /// todo: make them const
+            std::string group; /// permission
+            uint64_t version;  // API version ?

```

Comment:
Each SAL data class is versioned. This is not the SAL API version. This is defined in the datamodel.

Reply:
I cant comment on cppsal. This pull request is for the csal interface.

Reply:
so it is kind of API version, not data revision version.   I think we can fix it for one release for all data classes.

Reply:
Its the version of the data class itself. Each data class has a version number so the data class structures can be updated if a new attribute or modification is required e.g. Signal version 2 adds a new attribute. This allows a client to discriminate which version it is handling when returned by the server.

Old data stored in the storage array does not get promoted to the latest version (it is already written) so the client must be able to handle revisions of the data representation.

SAL is built to sit on top of object stores. The nature of the objects are allowed to evolve with use. Versioning is built into each and every level of the system to allow structured evolution and the ability to track the full history of changes.

The SAL REST API has a version
Each data classes have a version.
Each server subcomponent has a version.
The data tree is versioned (revisions)